Refactor remove let

USKTEA opened the pull request:
안녕하세요 리뷰어님 고생 많으십니다!
아직 배우고 있는 단계여서 과제의도에 맞게 작성했는지 모르겠네요. 
한 가지 여쭤보고 싶은 부분은 현재 render 함수 안에 handleClick 함수 선언했는데 이 함수를 밖으로 빼려고 
위에 배치하면 ESLint에서 render 함수를 선언하기 전에 사용했다고 경고가 뜨고, 아래에 배치하면 render 함수에서 
handleClick 함수가 선언되기 전에 호출되었다고 하는데 어떻게 해결하면 좋을까요?

Commit:
Revert "Refactor remove let"

This reverts commit 82f9444118fa5710a6bfaf7b0623cf94aea51fc8.

heyjyu commented:
> 한 가지 여쭤보고 싶은 부분은 현재 render 함수 안에 handleClick 함수 선언했는데 이 함수를 밖으로 빼려고
위에 배치하면 ESLint에서 render 함수를 선언하기 전에 사용했다고 경고가 뜨고, 아래에 배치하면 render 함수에서
handleClick 함수가 선언되기 전에 호출되었다고 하는데 어떻게 해결하면 좋을까요?
render함수 내부에 선언해보시는 건 어떨까요?? 그리고 scope에 대해서도 같이 공부해보시면 도움이 될 것 같습니다!

heyjyu commented on src/index.js:
```diff
@@ -1,2 +1,56 @@
+  return element;
+};
+
+const render = (countNumber = 0) => {
```
default값은 따로 선언을 해주시면 더 좋을 것 같아요!

heyjyu replied:
매개변수를 destructuring해주시면 나중에 2개 이상의 매개변수를 받을 때 더 의미를 명확하게 받을 수 있게 됩니다! 지금부터 연습을 해보시면 더 좋을 것 같아요!
참고: https://ko.javascript.info/destructuring-assignment#ref-1622

USKTEA replied:
> default값은 따로 선언을 해주시면 더 좋을 것 같아요!

default값을 따로 선언한다는 의미를 잘 모르겠습니다, 첫 render 함수가 호출될 때 default값인  0을 인자로 넣어 호출을 하라는 말씀이실까요?

heyjyu replied:
`const defaultValue = { count: 0 };`
와 같이 따로 선언을 해주시면 0의 의미를 조금 더 명확히 코드에 드러낼 수 있을 것 같습니다!

USKTEA replied:
넵 감사합니다!

heyjyu commented on src/index.js:
```diff
@@ -1,2 +1,56 @@
+};
+
+const render = (countNumber = 0) => {
+  const handleClick = (value) => {
```
한가지 함수는 한가지 기능만 하게 분리를 해주시면 더 좋을 것 같아요!

USKTEA replied:
````javascript 
const handleClickNumber = (value) => {
    render({ count: value });
  };

  const handleClickButton = () => {
    render({ count: count + 1 });
  };
````

일단 해당 형태로 분리하였습니다.  

제가 처음에 생각한건 비슷한 역할은 하는 button들이고 click의 콜백함수의 결과에 대해 기대하는 부분이 비슷하니 
조건을 달아서 조금 다르게 작동하게 했었던건데 , 결과 값이 다르다면 함수를 분리하는 것이 더 좋은 코드였나 보네요. 

혹시 수많은 버튼이 있고 수행하는 역할이 조금씩 다를 때는 그 만큼의 함수를 작성을 해야 되는 것인지 궁금합니다

heyjyu replied:
사람마다 다 판단이 다르고 꼭 제 말이 진리인 건 아니기도 합니다! 상황에 따라 판단을 해야될 것 같습니다! 의미적으로 handleClick이라고 함수를 명명하면 클릭이벤트를 다루는 함수라는 의미의 함수이기 때문에 기존에 작성하신 대로 놔둬도 의미가 파악이 잘 되긴 하지만, 카운트를 증가시킨다는 의미를 부여하고 싶다면, countUp같은 함수명을 사용해서 함수를 만들게 될텐데, 이렇게 조금 더 의미를 부여한 코드를 짜게 될때는 1, 2, 3을 클릭하는 건 countUp이 아니니 두 개를 분리하게 될 거예요! 결국에 코드가 어떤 스토리를 보여주고 싶냐에 따라 달려있는 것 같습니다!

USKTEA replied:
넵! 답변 감사합니다!

heyjyu commented on src/index.js:
```diff
@@ -1,2 +1,56 @@
+
+  const element = (
+    <div>
+      <h1 id="title">Assignment 1 : Remove var</h1>
```
굳~

heyjyu commented on src/index.js:
```diff
@@ -1,2 +1,56 @@
+    </div>
+  );
+
+  document.getElementById('app').textContent = '';
```
반복되는 부분은 따로 선언을 해주면 더 좋을 것 같아요!

USKTEA replied:
넵, 수정하겠습니다

Commit:
Refactor render function and handleClick function

Parameter를 구조분해할당 할 수 있도록 객체로 받게 수정.
반복적으로 선언되는 부분을 함수로 감싸 실행하도록 수정.
handleClick 함수의 기능을 분리함.

USKTEA commented:
> > 한 가지 여쭤보고 싶은 부분은 현재 render 함수 안에 handleClick 함수 선언했는데 이 함수를 밖으로 빼려고
> > 위에 배치하면 ESLint에서 render 함수를 선언하기 전에 사용했다고 경고가 뜨고, 아래에 배치하면 render 함수에서
> > handleClick 함수가 선언되기 전에 호출되었다고 하는데 어떻게 해결하면 좋을까요?
> 
> render함수 내부에 선언해보시는 건 어떨까요?? 그리고 scope에 대해서도 같이 공부해보시면 도움이 될 것 같습니다!

scope를 생각한다면  

````javascript 
const foo = () => {
bar();
};

const bar = () => {
console.log(1); 
};

foo();

````

1. cosnt foo가 전역 실행 컨텍스트 등록
2. const bar가 전역 실행 컨텍스트 등록
3. foo 호출 
4. bar는 foo 함수 컨텍스트에 존재하지 않아 스코프체인을 따라 상위 컨텍스트를 탐색 -> 전역 컨텍스트에서 bar 찾음
5. bar 호출

이렇게 되어서  오류는 발생하지 않는데 무엇 때문에 꼭 선언 후에 호출해야 한다는 것일까요?

heyjyu commented:
> > > 한 가지 여쭤보고 싶은 부분은 현재 render 함수 안에 handleClick 함수 선언했는데 이 함수를 밖으로 빼려고
> > > 위에 배치하면 ESLint에서 render 함수를 선언하기 전에 사용했다고 경고가 뜨고, 아래에 배치하면 render 함수에서
> > > handleClick 함수가 선언되기 전에 호출되었다고 하는데 어떻게 해결하면 좋을까요?
> > 
> > 
> > render함수 내부에 선언해보시는 건 어떨까요?? 그리고 scope에 대해서도 같이 공부해보시면 도움이 될 것 같습니다!
> 
> scope를 생각한다면
> 
> ```js
> const foo = () => {
> bar();
> };
> 
> const bar = () => {
> console.log(1); 
> };
> 
> foo();
> ```
> 
> 1. 변수 foo 전역 실행 컨텍스트 등록
> 2. 변수 bar 전역 실행 컨텍스트 등록
> 3. 변수 foo 값 할당
> 4. 변수 bar 값 할당
> 5. foo 호출
> 6. 변수 bar는 함수 foo의 컨텍스트에 존재하지 않아 스코프체인을 따라 상위 컨텍스트를 탐색
> 7. 전역 컨텍스트에서 변수 bar 찾음
> 8. bar 호출
> 
> 이렇게 되어서 오류는 발생하지 않을텐데 무엇 때문에 꼭 선언 후에 호출해야 한다는 것일까요?

https://eslint.org/docs/latest/rules/no-use-before-define를 참고해주시면 좋을 것 같습니다!

코드는 일단 돌아가는 것도 중요하지만, 보통 프로젝트를 여러명이 함께 하는 경우가 많기 때문에 작성은 나 혼자 한번 해도, 프로젝트 규모에 따라 읽히는 횟수는 어마어마해질 수 있습니다. 따라서 가독성이 좋은 코드를 써야 협업에 좋은데요, 이때 위에서 아래로 보통 코드를 읽기 때문에 선언을 하고, 사용을 하는 게 가독성이 더 좋기 때문이라고 생각해주시면 좋을 것 같습니다~

heyjyu commented on src/index.js:
```diff
@@ -1,2 +1,58 @@
+  const element = (
+    <div>
+      <h1 id="title">Assignment 1 : Remove let</h1>
+      <button type="button" onClick={() => handleClickButton()}>
```
여기는 값을 받지 않으니 () => 함수 꼴을 쓰지 않으면 더 깔끔한 코드를 작성하실 수 있을 것 같습니다!

USKTEA replied:
넵! 아래 값을 넘겨주는 함수가 있어 동일한 스타일을 유지하려고 해당 형태로 작성했는데, 
그러지 않고 넘기는 값이 없다면 함수를 바로 넣는 것이 더 깔끔한 코드군요

USKTEA commented:
> > > > 한 가지 여쭤보고 싶은 부분은 현재 render 함수 안에 handleClick 함수 선언했는데 이 함수를 밖으로 빼려고
> > > > 위에 배치하면 ESLint에서 render 함수를 선언하기 전에 사용했다고 경고가 뜨고, 아래에 배치하면 render 함수에서
> > > > handleClick 함수가 선언되기 전에 호출되었다고 하는데 어떻게 해결하면 좋을까요?
> > > 
> > > 
> > > render함수 내부에 선언해보시는 건 어떨까요?? 그리고 scope에 대해서도 같이 공부해보시면 도움이 될 것 같습니다!
> > 
> > 
> > scope를 생각한다면
> > ```js
> > const foo = () => {
> > bar();
> > };
> > 
> > const bar = () => {
> > console.log(1); 
> > };
> > 
> > foo();
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > 1. 변수 foo 전역 실행 컨텍스트 등록
> > 2. 변수 bar 전역 실행 컨텍스트 등록
> > 3. 변수 foo 값 할당
> > 4. 변수 bar 값 할당
> > 5. foo 호출
> > 6. 변수 bar는 함수 foo의 컨텍스트에 존재하지 않아 스코프체인을 따라 상위 컨텍스트를 탐색
> > 7. 전역 컨텍스트에서 변수 bar 찾음
> > 8. bar 호출
> > 
> > 이렇게 되어서 오류는 발생하지 않을텐데 무엇 때문에 꼭 선언 후에 호출해야 한다는 것일까요?
> 
> [https://eslint.org/docs/latest/rules/no-use-before-define를](https://eslint.org/docs/latest/rules/no-use-before-define%EB%A5%BC) 참고해주시면 좋을 것 같습니다!
> 
> 코드는 일단 돌아가는 것도 중요하지만, 보통 프로젝트를 여러명이 함께 하는 경우가 많기 때문에 작성은 나 혼자 한번 해도, 프로젝트 규모에 따라 읽히는 횟수는 어마어마해질 수 있습니다. 따라서 가독성이 좋은 코드를 써야 협업에 좋은데요, 이때 위에서 아래로 보통 코드를 읽기 때문에 선언을 하고, 사용을 하는 게 가독성이 더 좋기 때문이라고 생각해주시면 좋을 것 같습니다~

남겨주신 링크를 보니 확실히 이해가 되었습니다 감사합니다!

Commit:
Refactor click callback function name

Click callback function's name now have more specific name.
Add defaultvalue variable to render it at default.

heyjyu commented:
아주 잘하셨습니다! 회고 남겨주시면 머지하겠습니다~
